File: astar/RandomPuzzle.java
package astar;

import java.util.Arrays;
import java.util.Random;
import java.util.Map;
import java.util.HashMap;

class RandomPuzzle {
    private final int[][] basePuzzle;
    private final int size;

    public RandomPuzzle(int size) {
        this.basePuzzle = generatePuzzle(size);
        this.size = size;
    }

    private int[][] generatePuzzle(int size) {
        int[][] puz = new int[size][size];
        int count = 1;
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++) {
                puz[i][j] = count;
                count++;
            }
        }
        int row = size - 1;
        int col = size - 1;
        puz[row][col] = 0;
        return puz;
    }

    public int[][] getBasePuzzle() {
        return basePuzzle;
    }

    public int[][] getRandomisedPuzzle() {
        int[][] puzzle = new int[size][size];
        for (int i = 0; i < size; i++) {
            puzzle[i] = Arrays.copyOf(basePuzzle[i], size);
        }

        int row = size - 1;
        int col = size - 1;
        Random rand = new Random();
        int iter = rand.nextInt(100);
        Map<Integer, int[]> dirMap = new HashMap<>();
        dirMap.put(0, new int[] { -1, 0 });
        dirMap.put(1, new int[] { 1, 0 });
        dirMap.put(2, new int[] { 0, -1 });
        dirMap.put(3, new int[] { 0, 1 });

        int i = 0;
        while (i < iter) {
            int move = rand.nextInt(4);
            int[] dir = dirMap.get(move);

            int newRow = row + dir[0];
            int newCol = col + dir[1];

            if (newRow >= 0 && newRow < size && newCol >= 0 && newCol < size) {
                puzzle[row][col] = puzzle[newRow][newCol];
                puzzle[newRow][newCol] = 0;

                row = newRow;
                col = newCol;
                i++;
            }
        }

        return puzzle;
    }

}

File: astar/cpp/RandomPuzzle.cpp
#include "RandomPuzzle.h"
#include <random>

RandomPuzzle::RandomPuzzle(int size) : size(size) {
    basePuzzle = generatePuzzle(size);
}

std::vector<std::vector<int>> RandomPuzzle::generatePuzzle(int size) {
    std::vector<std::vector<int>> puz(size, std::vector<int>(size));
    int count = 1;
    for (int i = 0; i < size; i++) {
        for (int j = 0; j < size; j++) {
            puz[i][j] = count;
            count++;
        }
    }
    int row = size - 1;
    int col = size - 1;
    puz[row][col] = 0;
    return puz;
}

std::vector<std::vector<int>> RandomPuzzle::getBasePuzzle() const {
    return basePuzzle;
}

std::vector<std::vector<int>> RandomPuzzle::getRandomisedPuzzle() {
    std::vector<std::vector<int>> puzzle(size, std::vector<int>(size));
    for (int i = 0; i < size; i++) {
        for (int j = 0; j < size; j++) {
            puzzle[i][j] = basePuzzle[i][j];
        }
    }

    int row = size - 1;
    int col = size - 1;
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<int> iterDist(0, 9999);
    std::uniform_int_distribution<int> moveDist(0, 3);
    
    int iter = iterDist(gen);
    std::map<int, std::vector<int>> dirMap;
    dirMap[0] = {-1, 0};
    dirMap[1] = {1, 0};
    dirMap[2] = {0, -1};
    dirMap[3] = {0, 1};

    int i = 0;
    while (i < iter) {
        int move = moveDist(gen);
        std::vector<int> dir = dirMap[move];

        int newRow = row + dir[0];
        int newCol = col + dir[1];

        if (newRow >= 0 && newRow < size && newCol >= 0 && newCol < size) {
            puzzle[row][col] = puzzle[newRow][newCol];
            puzzle[newRow][newCol] = 0;

            row = newRow;
            col = newCol;
            i++;
        }
    }

    return puzzle;
}